### PR TITLE
 MULE-17065: Scatter gather and Parallel Foreach are not working inside error handler

### PR DESCRIPTION
**Pull request by mbuchwald**

### CODE REVIEW

File: integration/src/test/java/org/mule/test/integration/exceptions/TryAndErrorHandlingTestCase.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.CoreMatchers.is;
+import static org.mule.functional.api.component.InvocationCountMessageProcessor.getNumberOfInvocationsFor;
+
+import org.junit.Test;
+import org.mule.runtime.core.api.event.CoreEvent;
+import org.mule.test.AbstractIntegrationTestCase;
+
+
+public class TryAndErrorHandlingTestCase extends AbstractIntegrationTestCase {
```
Comment (elrodro83) [inline integration/src/test/java/org/mule/test/integration/exceptions/TryAndErrorHandlingTestCase.java]: allure

File: integration/src/test/java/org/mule/test/routing/ForeachTestCase.java [hunk 2]
```diff
@@ -593,4 +593,10 @@ public void errorAfterThreadChange() throws Exception {
     expectedException.expectCause(instanceOf(java.io.IOException.class));
     flowRunner("errorAfterThreadChange").withPayload(asList("0", "1", "2", "3", "4", "5", "6", "7", "8", "9")).run();
   }
+
+  @Test
```
Comment (elrodro83) [inline integration/src/test/java/org/mule/test/routing/ForeachTestCase.java]: allure

File: integration/src/test/java/org/mule/test/routing/ParallelForEachTestCase.java [hunk 3]
```diff
@@ -192,6 +193,12 @@ public void variables() throws Exception {
     flowRunner("variables").withPayload(fruitList).run();
   }
 
+  @Test
+  public void parallelForEachInErrorHandler() throws Exception {
```
Comment (elrodro83) [inline integration/src/test/java/org/mule/test/routing/ParallelForEachTestCase.java]: allure

File: integration/src/test/java/org/mule/test/routing/ScatterGatherRouterTestCase.java [hunk 4]
```diff
@@ -200,6 +200,18 @@ public void variables() throws Exception {
     flowRunner("variables").run();
   }
 
+  @Test
+  public void scatterGattherInsideErrorHandler() throws Exception {
```
Comment (elrodro83) [inline integration/src/test/java/org/mule/test/routing/ScatterGatherRouterTestCase.java]: allure

File: integration/src/test/java/org/mule/test/routing/ScatterGatherRouterTestCase.java [hunk 5]
```diff
@@ -200,6 +200,18 @@ public void variables() throws Exception {
...
+    CoreEvent event = flowRunner("scatterGattherInsideErrorHandler").run();
+    assertThat(event.getMessage().getPayload().getValue(), is("hello"));
+  }
+
+  @Test
+  public void scatterGattherInsideErrorHandlerThrowsError() throws Exception {
```
Comment (elrodro83) [inline integration/src/test/java/org/mule/test/routing/ScatterGatherRouterTestCase.java]: allure

File: integration/src/test/resources/org/mule/test/integration/exceptions/try-error-handling.xml [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="
+    http://www.mulesoft.org/schema/mule/core http://www.mulesoft.org/schema/mule/core/current/mule.xsd
+    http://www.mulesoft.org/schema/mule/test http://www.mulesoft.org/schema/mule/test/current/mule-test.xsd">
+
+    <flow name="tryInErrorHandler">
+        <raise-error type="APP:SOME"/>
+        <error-handler>
+            <on-error-continue>
+                <try>
+                    <logger message="hello"/>
```
Comment (elrodro83) [inline integration/src/test/resources/org/mule/test/integration/exceptions/try-error-handling.xml]: add a verification that this was actually executed

File: integration/src/test/resources/foreach-test.xml [hunk 7]
```diff
@@ -340,5 +340,16 @@
         </foreach>
     </flow>
 
+    <flow name="foreachInErrorHandler">
+        <raise-error type="APP:SOME"/>
+        <error-handler>
+            <on-error-continue>
+                <foreach collection="#[['apple']]">
+                    <logger></logger>
```
Comment (elrodro83) [inline integration/src/test/resources/foreach-test.xml]: add a verification that this was actually executed

File: integration/src/test/resources/parallel-foreach-config.xml [hunk 8]
```diff
@@ -220,4 +220,16 @@
         <test:assert expression="#[payload[2].payload == 'orange-result']"/>
     </sub-flow>
 
+    <flow name="parallelForEachInErrorHandler">
+        <raise-error type="APP:SOME"/>
+        <error-handler>
+            <on-error-continue>
+                <parallel-foreach collection="#[['apple']]">
+                    <logger message="Hello World"/>
```
Comment (elrodro83) [inline integration/src/test/resources/parallel-foreach-config.xml]: add a verification that this was actually executed

File: integration/src/test/resources/scatter-gather-test.xml [hunk 9]
```diff
@@ -204,6 +204,53 @@
         </error-handler>
     </flow>
 
+    <flow name="scatterGattherInsideErrorHandler">
+        <raise-error type="APP:SOME"/>
+        <error-handler>
+            <on-error-continue>
+                <scatter-gather>
+                    <route>
+                        <logger level="ERROR" message="Hello World"/>
```
Comment (elrodro83) [inline integration/src/test/resources/scatter-gather-test.xml]: add a verification that this was actually executed

File: integration/src/test/resources/scatter-gather-test.xml [hunk 10]
```diff
@@ -204,6 +204,53 @@
...
+                    </route>
+                    <route>
+                        <logger level="ERROR" message="Hello World"/>
```
Comment (elrodro83) [inline integration/src/test/resources/scatter-gather-test.xml]: add a verification that this was actually executed

File: integration/src/test/resources/scatter-gather-test.xml [hunk 11]
```diff
@@ -204,6 +204,53 @@
...
+    </flow>
+
+    <flow name="scatterGattherInsideErrorHandlerThrowsError">
+        <try>
+            <raise-error type="APP:SOME"/>
+            <error-handler>
+                <on-error-continue>
+                    <scatter-gather>
+                        <route>
+                            <logger level="ERROR" message="Hello World"/>
```
Comment (elrodro83) [inline integration/src/test/resources/scatter-gather-test.xml]: add a verification that this was actually executed

Comment (mbuchwald) [reply integration/src/test/resources/scatter-gather-test.xml]: It's down, in the error handler (it validates and also sets the payload that it's asserted in the test case)

**Pull request merged**